Add YamlNode.asJsonType()

tvolkert opened the pull request:
This will support use cases like the following:

```dart
@JsonSerializable()
class MyClass {
  ...
}

void main() {
  YamlMap yamlMap = loadYaml(content);
  MyClass obj = MyClass.fromJson(yamlMap.asJsonType());
}
```

tvolkert commented:
@nex3 @kevmoo

natebosch commented on lib/src/yaml_node.dart:
```diff
@@ -78,6 +81,18 @@ class YamlMap extends YamlNode with collection.MapMixin, UnmodifiableMapMixin {
 
+  @override
+  Map<String, dynamic> asJsonType() {
+    final Map<String, dynamic> result = <String, dynamic>{};
```
`final result = <String, dynamic>{};`

https://dart.dev/guides/language/effective-dart/design#avoid-type-annotating-initialized-local-variables

natebosch commented on lib/src/yaml_node.dart:
```diff
@@ -78,6 +81,18 @@ class YamlMap extends YamlNode with collection.MapMixin, UnmodifiableMapMixin {
+  @override
+  Map<String, dynamic> asJsonType() {
+    final Map<String, dynamic> result = <String, dynamic>{};
+    for (MapEntry<dynamic, YamlNode> entry in nodes.entries) {
```
`for (var entry in node.entries) {`

kevmoo commented:
Have you seen the support in `json_serializable` for handling Yaml sources directly?

https://github.com/dart-lang/json_serializable/blob/master/_test_yaml/test/src/build_config.dart

It comes with some very nice error reporting...

natebosch commented on lib/src/yaml_node.dart:
```diff
@@ -133,6 +148,10 @@ class YamlList extends YamlNode with collection.ListMixin {
 
+  @override
+  List<dynamic> asJsonType() =>
+      nodes.map((YamlNode node) => node.asJsonType()).toList();
```
https://dart.dev/guides/language/effective-dart/design#avoid-annotating-inferred-parameter-types-on-function-expressions

`nodes.map((node) => node.toJsonType()).toList();`

natebosch commented on lib/src/yaml_node.dart:
```diff
@@ -36,6 +36,9 @@ abstract class YamlNode {
   /// [List], respectively.
   get value;
+
+  /// Returns a copy of this YAML node as a JSON-compatible type.
```
If it is making a copy then the convention would be to name it `toJsonType`.

https://dart.dev/guides/language/effective-dart/design#prefer-naming-a-method-to___-if-it-copies-the-objects-state-to-a-new-object

We could also consider writing an adapter instead.

natebosch commented on lib/src/yaml_node_wrapper.dart:
```diff
@@ -41,6 +41,10 @@ class YamlMapWrapper extends MapBase
 
+  @override
+  Map<String, dynamic> asJsonType() =>
+      Map<String, dynamic>.from(_dartMap.cast<String, dynamic>());
```
Using both `Map.from` and `.cast` seems redundant... We should also be using inference here.

How about `Map.from(_dartMap);`

natebosch commented on lib/src/yaml_node_wrapper.dart:
```diff
@@ -111,6 +115,9 @@ class YamlListWrapper extends ListBase implements YamlList {
   }
 
+  @override
+  List<dynamic> asJsonType() => List<dynamic>.from(_dartList);
```
`_dartList.toList();`

kevmoo commented:
Tests would also be nice.

Also: why not call it `toJson`? It's a convention we use. Would make `jsonEncode(yaml)` "just work"

tvolkert commented:
> Have you seen the support in json_serializable for handling Yaml sources directly?

It didn't work for me...

```
type 'YamlMap' is not a subtype of type 'Map<String, dynamic>'
```

(the generated code  requires a `Map<String, dynamic>`

kevmoo commented:
> > Have you seen the support in json_serializable for handling Yaml sources directly?
> 
> It didn't work for me...
> 
> ```
> type 'YamlMap' is not a subtype of type 'Map<String, dynamic>'
> ```
> 
> (the generated code requires a `Map<String, dynamic>`)

Did you use `anyMap: true`?

tvolkert commented:
> We'll need a CHANGELOG entry and to bump the feature version in the pubspec.

> Tests would also be nice.

Yep - wanted to get high-level thoughts first.  I'll add a test and address comments and 

> > > Have you seen the support in json_serializable for handling Yaml sources directly?
> > 
> > 
> > It didn't work for me...
> > ```
> > type 'YamlMap' is not a subtype of type 'Map<String, dynamic>'
> > ```
> > 
> > 
> > (the generated code requires a `Map<String, dynamic>`)
> 
> Did you use `anyMap: true`?

Ah, that's the bit I was missing.  wdyt - is this change worth landing (after addressing review comments), or should I abandon it?

natebosch commented:
> is this change worth landing (after addressing review comments), or should I abandon it?

I think it would be a fine API to have as `toJson` - but if we don't know of a concrete immediate use case I don't see a reason to push on it. I'd drop if you're able to use existing solutions.

tvolkert commented:
Yep - no code is better than more code.   I'll drop this.  Thanks for the reviews!
